source/codegen/templates/service_helpers.mako

```diff
@@ -8,17 +8,28 @@
 <%
   config = data['config']
   output_parameters = [p for p in parameters if common_helpers.is_output_parameter(p)]
-  session_output_param = next((parameter for parameter in output_parameters if parameter['grpc_type'] == 'nidevice_grpc.Session'), None)
+  input_parameters = [p for p in parameters if common_helpers.is_input_parameter(p)]
+  session_output_param = next((p for p in output_parameters if p['grpc_type'] == 'nidevice_grpc.Session'), None)
+  session_behavior_param = next((p for p in input_parameters if p['grpc_type'] == 'nidevice_grpc.SessionInitializationBehavior'), None)
+  session_initialized_param = next((p for p in output_parameters if common_helpers.is_proto_only_parameter(p) and p['type'] == 'bool'), None)
   output_parameters_to_initialize = [p for p in output_parameters if p['grpc_type'] != 'nidevice_grpc.Session']
   resource_handle_type = session_output_param['type']
   session_output_var_name = common_helpers.get_cpp_local_name(session_output_param)
   close_function_call = function_data['custom_close'] if 'custom_close' in function_data else f"{config['close_function']}(id)"
 
   explicit_session_params = (common_helpers.get_cpp_local_name(param) for param in parameters if param.get('is_session_name', False))
   session_field_name = next(explicit_session_params, 'session_name')
+  add_session_snippet = 'add_session(grpc_device_session_name, init_lambda, cleanup_lambda, session_id)'
+  if session_behavior_param and session_initialized_param:
+      session_behavior_param_name = common_helpers.get_cpp_local_name(session_behavior_param)
+      session_initialized_param_name = common_helpers.get_cpp_local_name(session_initialized_param)
+      add_session_snippet = f'add_session(grpc_device_session_name, init_lambda, cleanup_lambda, session_id, {session_behavior_param_name}, &{session_initialized_param_name})'
 %>\
 ${initialize_input_params(function_name, parameters)}
 ${initialize_output_params(output_parameters_to_initialize)}\
+% if session_initialized_param:
+      bool ${session_initialized_param_name} = false;
```

Comment:
This is needed because `session_initialized_param_name` is not in `output_parameters_to_initialize`? Because `is_proto_only_parameter`? If there's a way to tweak the logic so that this goes in `output_parameters_to_initialize` (and perhaps tweak some names so the distinctions are clearer): I think that'd be better.

Reply:
It was in `output_parameters_to_initialize`, it was just that the `initialize_output_params` was filtering out proto_only params. I updated the `initialize_output_params` function to instead filter using a new method in common_helpers (get_params_needing_initialization) so that this new output param gets initialized the standard way through `initialize_output_params` now.